--- caffe-7d92d5c23d503af0565159b40c2d5155b5fdc3fb/CMakeLists.txt~	2016-08-14 03:48:55.000000000 +0000
+++ caffe-7d92d5c23d503af0565159b40c2d5155b5fdc3fb/CMakeLists.txt	2018-11-30 04:58:35.342786228 +0000
@@ -30,7 +30,7 @@
 caffe_option(USE_CUDNN "Build Caffe with cuDNN library support" ON IF NOT CPU_ONLY)
 caffe_option(BUILD_SHARED_LIBS "Build shared libraries" ON)
 caffe_option(BUILD_python "Build Python wrapper" ON)
-set(python_version "2" CACHE STRING "Specify which Python version to use")
+set(python_version "3" CACHE STRING "Specify which Python version to use")
 caffe_option(BUILD_matlab "Build Matlab wrapper" OFF IF UNIX OR APPLE)
 caffe_option(BUILD_docs   "Build documentation" ON IF UNIX OR APPLE)
 caffe_option(BUILD_python_layer "Build the Caffe Python layer" ON)
